### PR TITLE
feat: add council-review-action composite GitHub Action

### LINKED ISSUE #253: feat: council-review-action — reusable GitHub Action for CI council reviews
## Context

The `complytime/org-infra` project is prototyping a CI-based council review workflow
that uses OpenCode to run each Divisor persona, with agent definitions read dynamically
from the repo's `.opencode/agents/divisor-*.md` files. The pattern is implemented as a
custom three-workflow chain (collect → consumer → reusable) specific to that org.

Design decisions and rationale are documented in:
`openspec/changes/council-review-orchestration/design.md` (in `complytime/org-infra`)

**org-infra is the prototype** (design.md D8). The generic orchestration logic will be
extracted into this action after validation. org-infra then becomes a thin consumer
handling org-specific concerns (WIF auth, fork-safe chain) while calling this action
for the review itself.

## What to extract from the prototype

The following logic is generic and belongs in this action, not in org-infra:

- **Agent discovery**: Glob `.opencode/agents/divisor-*.md`, detect repo-local
  or use bundled fallback from the action
- **Prompt composition**: Minimal CI instruction referencing `review-council.md`,
  `review-pr.md`, severity packs, and convention packs as methodology
- **Pre-fetch pattern**: Collect CI checks, existing reviews, linked issues as JSON
  files for the model to read (no Shell access in CI)
- **`opencode run` invocation**: With Vertex AI provider, prompt file, model selection
  (changed from `claude -p --agents` per ADR-001)
- **Output parsing**: JSON with `summary` + `inline_comments` (path, line, body
  with severity and persona attribution)
- **Diff size guardrails**: Hard limit for massive diffs, adaptive budget based on
  repo PR history, noise file filtering, spec file allowance
- **Fallback**: Single-agent mode when no `divisor-*.md` files exist, comment mode
  when JSON parsing fails, skip mode when diff exceeds hard limit

## What stays in the consuming org

- Auth provider configuration (Vertex AI WIF, Bedrock, direct API key)
- Fork-safe workflow chain (collect → consumer → reusable)
- Secret forwarding and concurrency control
- PR comment posting with bot marker deduplication
- Old review cleanup (supersede stale bot reviews)

## Acceptance criteria

- [x] Composite action extracts review logic from org-infra prototype
- [x] Agent discovery works (repo-local and bundled fallback)
- [x] Pre-fetch pattern covers review-pr.md methodology
- [x] Structured JSON output (summary + inline_comments) parsed correctly
- [x] Inline comments filtered to diff-visible lines
- [x] Fallback modes: single-agent, comment-only, review-skipped
- [x] Diff size guardrails: hard limit (10K), adaptive budget, noise filtering
- [x] End-to-end evidence: complytime/org-infra#425
- [ ] Pinned to a release SHA (not feature branch) before downstream distribution
- [ ] At least 10 real PR reviews run successfully

## Also considered (lower priority)

- **Container image**: Less compelling since agent definitions are read at runtime —
  pre-baked agents would be stale after `uf init` updates.
- **Full `/review-council` parity**: Fix loops, local tool execution, Gaze analysis.
  Requires Shell access on untrusted diff content — security risk in CI.
- **BYOK externalization**: Tracked separately in #330.

### PR DESCRIPTION
**Pull request by sonupreetam**
## Summary

Adds a composite GitHub Action (`council-review-action/`) for AI-assisted
code review using OpenCode on Vertex AI. Extracted from the
complytime/org-infra prototype (PR complytime/org-infra#428), as specified
in #253.

**What the action does:**
- Auto-discovers Divisor persona definitions from `.opencode/agents/divisor-*.md`
  (repo-local or bundled fallback)
- Pre-fetches PR context (CI checks, existing reviews, linked issues) as JSON
- Filters noise files from diff (lock files, vendor, generated code)
- Annotates diff with `[L<N>]` source-file line numbers for accurate inline comments
- Builds a constrained review prompt delegating methodology to
  `.opencode/commands/review-council.md` and `review-pr.md`
- Invokes `opencode run` with `google-vertex-anthropic` provider for Claude
  on Vertex AI via WIF
- Parses structured JSON output (summary + inline comments) from OpenCode
  JSONL streaming format
- Filters inline comments to diff-visible lines (prevents GitHub API 422)
- Rescues out-of-diff findings into the summary body
- Falls back gracefully: single-agent mode when no personas exist, comment
  mode when JSON parsing fails

**Supply-chain safety:**
- OpenCode pinned to `opencode-ai@1.2.26` via npm
- All GitHub Actions in the caller pinned to full SHA

Includes documentation: README (architecture), docs/decisions.md (key design
decisions), docs/testing.md (coverage matrix), and test/test-pipeline.sh
(39 assertions across 15 test scenarios).

Replaces #329 (closed: cluttered timeline from development iterations).

Closes #253

## Evidence

End-to-end validation in [complytime/org-infra#427](https://github.com/complytime/org-infra/pull/427):
- Council review posted 8 inline comments on valid source-file lines (8/8)
- Line annotation approach eliminated patch-position confusion
- Full workflow chain: collect → consumer → reusable → action → inline review

## Test plan

- [x] WIF auth from `unbound-force` org validated (PR #235)
- [x] WIF auth from `complytime` org validated (PR complytime/org-infra#313)
- [x] Inline comments posted on correct diff lines (8/8 on #427)
- [x] Noise files excluded from diff before review
- [x] Fallback to PR comment when structured JSON parsing fails
- [x] Pinned `opencode-ai@1.2.26` install validated end-to-end
- [x] Local pipeline tests: 15 scenarios, 39 assertions passing

## Post-merge checklist

- [ ] Update action SHA in `complytime/org-infra` `reusable_council_review.yml` to merged `main` commit
- [ ] Switch `ci_council_review.yml` `uses:` from `@feat/council-review-production` to `@main`
- [ ] Verify full automatic chain fires on a real PR
- [ ] Delete stale branches: `feat/council-review-action`, `feat/council-review-action-v2`

### CODE REVIEW

File: council-review-action/action.yml [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+          echo "installed=true" >> "$GITHUB_OUTPUT"
+        fi
+
+    - name: Filter diff noise
+      if: steps.install.outputs.installed == 'true'
+      shell: bash
+      env:
+        DIFF_PATH: ${{ inputs.diff-path }}
+      run: |
+        SCRIPT="${{ github.action_path }}/scripts"
```
Comment (yvonnedevlinrh) [inline council-review-action/action.yml]: HIGH
${{ github.action_path }} interpolated directly in run: blocks instead of via env:, inconsistent with how inputs.* values are handled in the same file. Not an active injection vector (value is GitHub-controlled) but establishes a copy-paste pattern that is dangerous when applied to user-controllable expressions.

File: council-review-action/action.yml [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        bash "${SCRIPT}/prepare-diff.sh"
+
+    - name: Pre-fetch PR context
+      if: steps.install.outputs.installed == 'true'
+      shell: bash
+      env:
+        GH_TOKEN: ${{ inputs.github-token }}
+        META_PATH: ${{ inputs.meta-path }}
+      run: |
+        SCRIPT="${{ github.action_path }}/scripts"
```
Comment (yvonnedevlinrh) [inline council-review-action/action.yml]: Same as above

File: council-review-action/action.yml [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+          fi
+        fi
+
+    - name: Build review prompt
+      if: steps.install.outputs.installed == 'true'
+      shell: bash
+      env:
+        META_PATH: ${{ inputs.meta-path }}
+      run: |
+        SCRIPT="${{ github.action_path }}/scripts"
```
Comment (yvonnedevlinrh) [inline council-review-action/action.yml]: same as above

File: council-review-action/action.yml [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        bash "${SCRIPT}/build-prompt.sh"
+
+    - name: Run council review
+      if: steps.install.outputs.installed == 'true'
+      id: review
+      shell: bash
+      env:
+        MODEL: ${{ inputs.model }}
+      run: |
+        SCRIPT="${{ github.action_path }}/scripts"
```
Comment (yvonnedevlinrh) [inline council-review-action/action.yml]: Same as above

File: council-review-action/action.yml [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            > review_output.json
+          echo "review_json=review_output.json" \
+            >> "$GITHUB_OUTPUT"
+          echo "review_mode=comment" >> "$GITHUB_OUTPUT"
+          echo "::notice::Review skipped — no output file"
+          exit 0
+        fi
+
+        KEYS='.summary and .inline_comments'
+        SCRIPT="${{ github.action_path }}/scripts"
```
Comment (yvonnedevlinrh) [inline council-review-action/action.yml]: same as above

File: council-review-action/scripts/prepare-diff.sh [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+# NOTE: Spec/doc files (openspec/, .specify/, docs/) are NOT excluded.
+# Divisor personas specifically review specs for intent drift,
+# constitution alignment, and completeness.
+#
+# Required env: DIFF_PATH
+set -euo pipefail
+
+EXCLUDE_PATTERNS=(
+  # Lock files
+  'go\.sum'
```
Comment (yvonnedevlinrh) [inline council-review-action/scripts/prepare-diff.sh]: MEDIUM
Unanchored exclude regex: go\.sum matches substring go.sum inside filenames like go.summary. All patterns have the same issue — vendor/ matches event-vendor/, testdata/ matches sometestdata/, etc. Real files can be silently excluded from review.

File: council-review-action/action.yml [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+        # Case 1: raw output is already structured JSON
+        if jq -e "${KEYS}" review_raw.txt \
+          > /dev/null 2>&1; then
+          cp review_raw.txt review_parsed.json
+        else
+          # Case 2: streaming JSONL from opencode run --format json.
+          grep '"type":"text"' review_raw.txt \
```
Comment (yvonnedevlinrh) [inline council-review-action/action.yml]: MEDIUM
JSONL grep '"type":"text"' requires compact JSON with no whitespace. If OpenCode emits "type": "text" (with space), structured inline comments silently degrade to comment-only fallback with no error — only a ::notice::

File: openspec/changes/council-review-action/specs/council-review-action/spec.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+The action SHALL discover Divisor persona definitions by globbing
+`.opencode/agents/divisor-*.md` in the checked-out repository. It
+SHALL construct `--agents` JSON from the discovered files.
+
+#### Scenario: Multiple personas discovered
+
+- **WHEN** the repo contains 5+ `divisor-*.md` files
+- **THEN** the action builds an `--agents` JSON object with one
+  entry per persona and invokes `claude -p --agents`
```
Comment (yvonnedevlinrh) [inline openspec/changes/council-review-action/specs/council-review-action/spec.md]: MEDIUM
Spec scenarios reference claude -p --agents and --allowedTools. Tasks reference discover-agents.py and claude -p. Implementation uses opencode run. Spec artifacts were not updated after ADR-001 migration

File: openspec/changes/council-review-action/specs/council-review-action/spec.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+#### Scenario: Zero personas discovered
+
+- **WHEN** the repo contains no `divisor-*.md` files
+- **THEN** the action falls back to single-agent mode, logs a
+  `::notice::`, and invokes `claude -p` without `--agents`
```
Comment (yvonnedevlinrh) [inline openspec/changes/council-review-action/specs/council-review-action/spec.md]: Same as above

File: openspec/changes/council-review-action/tasks.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+- [x] 2.1 Create `prefetch.sh` — fetch CI check results via
+  `gh pr checks` → `pr-checks.json`
+- [x] 2.2 Fetch existing reviews via `gh api` → `pr-reviews.json`
+  and inline comments → `pr-review-comments.json`
+- [x] 2.3 Resolve linked issues from PR body (Fixes/Closes/Resolves
+  #N) → `pr-linked-issues.json`, limit 5 issues, truncate bodies
+
+## 3. Agent Discovery and Prompt Construction
+
+- [x] 3.1 Create `discover-agents.py` — glob `divisor-*.md`, build
```
Comment (yvonnedevlinrh) [inline openspec/changes/council-review-action/tasks.md]: Spec scenarios reference claude -p --agents and --allowedTools. Tasks reference discover-agents.py and claude -p. Implementation uses opencode run.

File: openspec/changes/council-review-action/tasks.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  `--agents` JSON via `json.dumps()`, output to file
+- [x] 3.2 Create `build-prompt.sh` — construct review prompt from
+  methodology file references (review-council.md, review-pr.md,
+  severity.md, convention packs), CI constraints, JSON output schema
+- [x] 3.3 Implement fallback flag when zero agents discovered
+
+## 4. Claude Invocation and Output Parsing
+
+- [x] 4.1 Create `run-review.sh` — invoke `claude -p` with
```
Comment (yvonnedevlinrh) [inline openspec/changes/council-review-action/tasks.md]: Spec scenarios reference claude -p --agents and --allowedTools. Tasks reference discover-agents.py and claude -p. Implementation uses opencode run.

File: council-review-action/scripts/run-review.sh [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+if [[ "${PROVIDER}" == "google-vertex-anthropic" ]]; then
+  export OPENCODE_CONFIG_CONTENT
+  OPENCODE_CONFIG_CONTENT=$(cat <<OCEOF
+{
+  "\$schema": "https://opencode.ai/config.json",
+  "provider": {
+    "google-vertex-anthropic": {
+      "models": {
+        "${MODEL_NAME}": {}
```
Comment (yvonnedevlinrh) [inline council-review-action/scripts/run-review.sh]: MEDIUM
${MODEL_NAME} interpolated into JSON heredoc without escaping. A model input containing " produces malformed JSON. Fix: construct with jq -n --arg.

File: council-review-action/scripts/run-review.sh [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+#
+# Run council review via opencode run. OpenCode auto-discovers
+# .opencode/ context (agents, commands, packs) and delegates to
+# Divisor personas when available.
+#
+# Required env: MODEL
+# Optional env: GOOGLE_CLOUD_PROJECT, VERTEX_LOCATION,
+#               GOOGLE_APPLICATION_CREDENTIALS
+# TODO: restore set -euo pipefail once OpenCode+Vertex is stable
+set -uo pipefail
```
Comment (yvonnedevlinrh) [inline council-review-action/scripts/run-review.sh]: MEDIUM
set -e omitted (set -uo pipefail only). The opencode run failure is captured, but any other command failure silently continues. TODO comment acknowledges this.

File: council-review-action/scripts/prefetch.sh [hunk 14]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: Apache-2.0
```
Comment (yvonnedevlinrh) [inline council-review-action/scripts/prefetch.sh:2]: MEDIUM
No test coverage for this script. test-pipeline.sh covers the four other scripts but not this one.

File: council-review-action/scripts/run-review.sh [hunk 15]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
```
Comment (yvonnedevlinrh) [inline council-review-action/scripts/run-review.sh:1]: MEDIUM
No test coverage for this script. test-pipeline.sh covers the four other scripts but not this one.

File: council-review-action/action.yml (hunk 1 above)
Comment (jflowers) [inline council-review-action/action.yml]: **SEC-001 [HIGH]**: `${{ github.action_path }}` is interpolated directly into this `run:` block via raw string substitution before bash executes. This is an expression injection risk — if the path contains shell metacharacters, it becomes a shell injection vector.

This pattern appears in 5 steps, but the "Detect Divisor agents" step (line 89) already correctly uses `env:` for `ACTION_PATH`. Apply the same safe pattern here and in the other 4 occurrences:

```suggestion
      env:
        ACTION_PATH: ${{ github.action_path }}
      run: |
        SCRIPT="${ACTION_PATH}/scripts"
```

File: council-review-action/action.yml [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+
+    - name: Detect Divisor agents
+      if: steps.install.outputs.installed == 'true'
+      id: agents
+      shell: bash
+      env:
+        AGENTS_PATTERN: ${{ inputs.agents-pattern }}
+        ACTION_PATH: ${{ github.action_path }}
+      run: |
+        COUNT=$(find . -path "./${AGENTS_PATTERN}" \
```
Comment (jflowers) [inline council-review-action/action.yml:105]: **SEC-002 [HIGH]**: `agents-pattern` is a user-configurable input passed to `find -path` without validation. A malicious caller could set it to a path traversal pattern like `../../etc/` or patterns with shell metacharacters.

The double-quoting around `${AGENTS_PATTERN}` prevents command injection, but `find` will happily traverse outside the repo.

**Recommendation**: Add input validation at the start of this step:
```bash
if [[ "${AGENTS_PATTERN}" == *..* ]] || [[ "${AGENTS_PATTERN}" == /* ]]; then
  echo "::error::agents-pattern must not contain '..' or absolute paths"
  exit 1
fi
if ! [[ "${AGENTS_PATTERN}" =~ ^[a-zA-Z0-9_./*-]+$ ]]; then
  echo "::error::agents-pattern contains invalid characters"
  exit 1
fi
```

File: council-review-action/scripts/run-review.sh (hunk 13 above)
Comment (jflowers) [inline council-review-action/scripts/run-review.sh]: **SEC-003 [HIGH]**: Missing `set -e` means commands before the explicit `|| OPENCODE_EXIT=$?` on line 39 can fail silently. The `PROVIDER` parameter expansion, `MODEL_NAME` extraction, and `OPENCODE_CONFIG_CONTENT` heredoc construction would all continue on failure.

The OpenCode invocation already captures its exit code explicitly (`|| OPENCODE_EXIT=$?`), so restoring `-e` won't affect that line.

**Recommendation**: Restore `set -euo pipefail` and remove the TODO.

File: council-review-action/scripts/run-review.sh (hunk 15 above)
Comment (jflowers) [inline council-review-action/scripts/run-review.sh:1]: **SEC-004 [HIGH]**: This script inherits the full environment from the composite action, including `GH_TOKEN` and WIF credentials. The security risk register documents this (A3) with `OPENCODE_PERMISSION` (#337) as future mitigation — but it's not enforced today.

**Recommendation**: At minimum, unset sensitive environment variables before the `opencode run` invocation:
```bash
unset GH_TOKEN GOOGLE_APPLICATION_CREDENTIALS 2>/dev/null || true
opencode run ...
```

Also consider adding `--pure` if OpenCode supports it to prevent external plugin loading.

File: openspec/changes/council-review-action/specs/council-review-action/spec.md (hunk 8 above)
Comment (jflowers) [inline openspec/changes/council-review-action/specs/council-review-action/spec.md]: **CRITICAL — Spec/implementation divergence**: This scenario references `--agents` JSON construction and `claude -p --agents`, but ADR-001 chose OpenCode and the implementation uses `opencode run` with native agent discovery. The same issue appears throughout `proposal.md`, `tasks.md`, and `design.md`.

The spec artifacts describe a fundamentally different system than what was built. Future readers will form an incorrect mental model.

**Recommendation**: Update all spec artifacts to reference `opencode run` with native `.opencode/agents/` discovery. Replace `claude -p` → `opencode run`, remove `--agents` JSON construction references, remove `discover-agents.py` references.

File: openspec/changes/council-review-action/.openspec.yaml [hunk 17]
```diff
@@ -0,0 +1,3 @@
+change: council-review-action
+schema: unbound-force
+status: proposed
```
Comment (jflowers) [inline openspec/changes/council-review-action/.openspec.yaml]: **HIGH — Stale status**: Status is `proposed` but all tasks in sections 1-4 are marked `[x]` and implementation is complete. This is a phase discipline concern.

**Recommendation**: Update to `status: implemented` (or `in-review`). Also mark validation tasks 5.1-5.3 as `[x]` in `tasks.md` since the test suite covers them.

File: council-review-action/README.md [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+│   ├── build-prompt.sh     # Prompt construction
+│   ├── run-review.sh       # OpenCode invocation
+│   ├── prefetch.sh         # PR context pre-fetch (CI, reviews)
+│   ├── extract-review-json.py  # JSON extraction from JSONL
+│   └── filter-diff-lines.py    # Line number validation
+├── test/
+│   └── test-pipeline.sh    # Pipeline tests (39 assertions)
+└── docs/
+    ├── decisions.md         # Key technical decisions
+    └── testing.md           # Test coverage and strategy
```
Comment (jflowers) [inline council-review-action/README.md:130]: **HIGH — Missing file in directory listing**: `docs/security-risks.md` (142 lines) is part of this PR but is omitted from this directory listing.

```suggestion
└── docs/
    ├── decisions.md         # Key technical decisions
    ├── security-risks.md    # Security risk register
    └── testing.md           # Test coverage and strategy
```

File: openspec/changes/council-review-action/design.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+(the Claude Code CLI approach).
+
+**Why rejected**: See ADR-001. OpenCode is the team standard, auto-
+discovers agent definitions natively, and eliminates the Python
+JSON construction pipeline. Agent frontmatter (permissions, mode)
+is used directly without translation.
+
+### D3: Diff in file, not in prompt string
+
+**Decision**: The diff stays in `pr-diff-truncated.patch` on disk.
```
Comment (jflowers) [inline openspec/changes/council-review-action/design.md]: **HIGH — Stale filename**: References `pr-diff-truncated.patch` which doesn't exist in the implementation. The actual pipeline produces `pr-diff-filtered.patch` (noise-filtered) and `pr-diff-annotated.patch` (line-annotated). There is no truncation step.

**Recommendation**: Update to reference `pr-diff-filtered.patch` and `pr-diff-annotated.patch`.

**Review by jflowers — CHANGES_REQUESTED** [review]
## Review: PR #336 — `feat: add council-review-action composite GitHub Action`

**Reviewer**: `/review-pr` (5 Divisor agents: Adversary, Architect, Guard, Testing, Scribe)
**CI**: All checks passing | **Existing reviews**: None

---

### Walkthrough

This PR introduces `council-review-action/`, a composite GitHub Action for AI-assisted code review using OpenCode on Vertex AI. The action auto-discovers Divisor personas, pre-fetches PR context (CI checks, reviews, linked issues), filters noise from diffs, annotates lines, builds a structured review prompt, invokes `opencode run`, and parses JSON output with diff-line validation. Includes 6 scripts, a 39-assertion test suite, comprehensive docs, an ADR, and OpenSpec artifacts.

---

### Consolidated Verdict: REQUEST CHANGES

All five reviewers independently reached REQUEST CHANGES. The implementation quality is strong — the scripts are well-separated, the test suite is thoughtful, the security risk register is exemplary, and the defensive prompt construction shows mature engineering. However, several systemic issues require resolution.

---

### Blocking Findings

| # | Sev | Finding | Files |
|---|-----|---------|-------|
| 1 | CRITICAL | Spec artifacts describe Claude CLI (`claude -p --agents`) despite ADR choosing OpenCode | spec.md, proposal.md, tasks.md, design.md |
| 2 | HIGH | `${{ github.action_path }}` expression injection in 5 `run:` blocks (step "Detect Divisor agents" already uses `env:` correctly) | action.yml |
| 3 | HIGH | `agents-pattern` input used in `find -path` without validation — path traversal risk | action.yml |
| 4 | HIGH | OpenCode process inherits `GH_TOKEN`/WIF credentials with no enforcement | action.yml, run-review.sh |
| 5 | HIGH | Missing `set -e` in `run-review.sh` — errors silently swallowed | run-review.sh |
| 6 | HIGH | File permissions: `prepare-diff.sh` and `filter-diff-lines.py` are 100644 (not executable) while all other scripts are 100755 | scripts/ |
| 7 | HIGH | AGENTS.md not updated with `council-review-action/` and `docs/decisions/` directories; no website documentation issue filed | AGENTS.md |
| 8 | HIGH | README directory listing omits `security-risks.md` | README.md |
| 9 | HIGH | `.openspec.yaml` status is `proposed` despite implementation being complete; validation tasks 5.1-5.3 unchecked | .openspec.yaml, tasks.md |

### Non-blocking Findings (recommended)

| Sev | Finding |
|-----|---------|
| MED | `npm install -g opencode-ai@1.2.26` without integrity verification |
| MED | Prompt injection surface via PR title/metadata (well-documented accepted risk) |
| MED | `extract-review-json.py` quadratic parsing on adversarial input — add size/iteration limits |
| MED | ~80-line parse-output logic inline in action.yml — extract to testable script |
| MED | `run-review.sh` and `prefetch.sh` have zero tests despite mockable boundaries |
| MED | Test suite not wired into CI — no automated regression protection |
| MED | Missing edge cases: empty diff, multiple JSON objects, assertion depth on filter-diff-lines |
| MED | Prompt structure tests don't verify JSON schema fields |
| MED | ADR in `docs/decisions/` creates ambiguity with `council-review-action/docs/decisions.md` |
| MED | Security risk T2 should lead with "no enforcement" rather than burying it |
| LOW | Silent API fallbacks in prefetch.sh without `::warning::` |
| LOW | `prepare-diff.sh` holds entire diff in shell variable — consider piping to file directly |

### Positive Observations

- Security risk register (`docs/security-risks.md`) is exemplary — honest about residual risks, structured T/A classification
- Test isolation well-implemented (`mktemp -d` + trap cleanup + subshells)
- Script separation of concerns is clean — each script handles one pipeline stage
- Defensive prompt construction with injection preamble and JSON schema constraint
- Multi-tier agent fallback chain (repo → bundled → single-agent) is well-designed
- Supply chain: all GitHub Actions pinned to full SHA

---

*Review generated by `/review-pr` with 5 Divisor agents (Adversary, Architect, Guard, Testing, Scribe)*

**Review by yvonnedevlinrh — CHANGES_REQUESTED** [review]

**Comment by sonupreetam** [comment]
Thank you for your review comments and feedbacks, it seems I didn't run review council on this old PR. I have taken care of all the review feedbacks. However, I will be thoroughly testing the outcome of the changes in the review outputs and moving this to draft until its ready.

File: council-review-action/scripts/run-review.sh (hunk 15 above)
Comment (sonupreetam) [reply council-review-action/scripts/run-review.sh:1]: Partially applied — unsetting `GH_TOKEN` is correct (not needed by the review invocation), but unsetting `GOOGLE_APPLICATION_CREDENTIALS` broke Vertex AI auth.

The `google-github-actions/auth` WIF step sets `GOOGLE_APPLICATION_CREDENTIALS` to point to the short-lived credential file. The Vertex AI SDK reads this env var to authenticate API calls. Unsetting it caused OpenCode to fail immediately (~3s instead of 30-120s) with a non-JSON auth error, silently falling back to comment mode.

Evidence: [complytime/org-infra run #30445719456](https://github.com/complytime/org-infra/actions/runs/30445719456) — review step completed in 3182ms with `Non-JSON output — comment fallback`.

Fixed in 48c9d01: only unset `GH_TOKEN`, keep `GOOGLE_APPLICATION_CREDENTIALS` intact. The `--pure` flag (tracked in #337) remains the proper long-term solution for sandboxing credentials.

**Comment by sonupreetam** [comment]
@jflowers & @yvonnedevlinrh The PR is ready for review again, the complytime/org-infra#427 is the evidence of the AI review on the PR.

File: council-review-action/scripts/prefetch.sh (hunk 14 above)
Comment (sonupreetam) [reply council-review-action/scripts/prefetch.sh:2]: Acknowledged — both `prefetch.sh` and `run-review.sh` are intentionally excluded from local tests because they make live API calls (`gh api`, `opencode run` on Vertex AI) that can't be meaningfully tested without mocking the entire CLI.

The testing strategy (`docs/testing.md`) documents this explicitly and describes the evidence PR process for integration validation: push to org-infra, trigger the consumer workflow, verify inline comments land on correct diff lines.

The 60-assertion local test suite covers the other 5 scripts (prepare-diff, build-prompt, extract-review-json, filter-diff-lines, parse-output) including all fallback paths that `parse-output.sh` exercises when `run-review.sh` output is malformed or missing.

File: council-review-action/scripts/run-review.sh (hunk 15 above)
Comment (sonupreetam) [reply council-review-action/scripts/run-review.sh:1]: Same as above — `run-review.sh` requires OpenCode + Vertex AI credentials to execute. Validated via evidence PRs (see `docs/testing.md`). The downstream parse pipeline (`parse-output.sh`) is fully tested for all output scenarios this script can produce: direct JSON, JSONL streaming, plain text, empty output, and timeout.

**Review by yvonnedevlinrh — APPROVED** [review]
LGTM, any findings are low or not an issue with this functionality

**Review by jflowers — APPROVED** [review]
All 10 blocking findings from the prior CHANGES_REQUESTED review (commit eece8d4) have been addressed: expression injection fixed, agents-pattern validation added, set -euo pipefail restored, credential scoping implemented, spec artifacts updated, file permissions fixed, AGENTS.md updated, README directory listing completed, OpenSpec status set to in-review, parse-output logic extracted to testable script.

Remaining LOW findings (non-blocking): npm install without content hash verification (npm's built-in integrity check covers this), GOOGLE_APPLICATION_CREDENTIALS in environment (documented trade-off, tracked in #337).

Implementation quality is strong — the security risk register is exemplary, the test suite (60 assertions) provides meaningful coverage, and the separation of concerns across scripts is clean.

_This review was generated by /review-pr (AI-assisted)._
